### ISSUE TITLE
Error when try to connect through Apache Spark

### ISSUE THREAD

**Issue by seufagner** [yandex/clickhouse-jdbc#109]
When I try to load some table using the current driver  version (builded with _mvn package_) through pyspark I got an error above:

```
def load_table(table_name):
    return spark.read.format("jdbc").options(
        driver=JDBC_DRIVER,
        url=CONNECTION_URL,
        dbtable=table_name
    ).load()
````

```
17/06/22 18:38:08 INFO clickhouse.ClickHouseDriver: Driver registered
17/06/22 18:38:08 INFO clickhouse.ClickHouseDriver: Creating connection

........

py4j.protocol.Py4JJavaError: An error occurred while calling o47.load.
: java.lang.NoSuchFieldError: INSTANCE
	at org.apache.http.impl.io.DefaultHttpRequestWriterFactory.<init>(DefaultHttpRequestWriterFactory.java:52)
	at org.apache.http.impl.io.DefaultHttpRequestWriterFactory.<init>(DefaultHttpRequestWriterFactory.java:56)
	at org.apache.http.impl.io.DefaultHttpRequestWriterFactory.<clinit>(DefaultHttpRequestWriterFactory.java:46)
	at org.apache.http.impl.conn.ManagedHttpClientConnectionFactory.<init>(ManagedHttpClientConnectionFactory.java:82)
	at org.apache.http.impl.conn.ManagedHttpClientConnectionFactory.<init>(ManagedHttpClientConnectionFactory.java:95)
	at org.apache.http.impl.conn.ManagedHttpClientConnectionFactory.<init>(ManagedHttpClientConnectionFactory.java:104)
	at org.apache.http.impl.conn.ManagedHttpClientConnectionFactory.<clinit>(ManagedHttpClientConnectionFactory.java:62)
	at org.apache.http.impl.conn.PoolingHttpClientConnectionManager$InternalConnectionFactory.<init>(PoolingHttpClientConnectionManager.java:572)
	at org.apache.http.impl.conn.PoolingHttpClientConnectionManager.<init>(PoolingHttpClientConnectionManager.java:174)
	at org.apache.http.impl.conn.PoolingHttpClientConnectionManager.<init>(PoolingHttpClientConnectionManager.java:158)
	at ru.yandex.clickhouse.util.ClickHouseHttpClientBuilder.getConnectionManager(ClickHouseHttpClientBuilder.java:77)
	at ru.yandex.clickhouse.util.ClickHouseHttpClientBuilder.buildClient(ClickHouseHttpClientBuilder.java:54)
	at ru.yandex.clickhouse.ClickHouseConnectionImpl.<init>(ClickHouseConnectionImpl.java:69)
	at ru.yandex.clickhouse.ClickHouseDriver.connect(ClickHouseDriver.java:58)
	at ru.yandex.clickhouse.ClickHouseDriver.connect(ClickHouseDriver.java:50)
	at ru.yandex.clickhouse.ClickHouseDriver.connect(ClickHouseDriver.java:32)
```

**Comment by serebrserg** [comment]
As I understand the problem there is likely a problem with classpath. 
The problem is likely to be that the driver has dependency on apache http client, and spark (or your application) has its own dependency on different version of http client, that is why DefaultHttpRequestWriterFactory according to your error is different from what I see in it from driver's dependencies.
I'm not familiar with pyspark, but I guess you could try to use --driver-class-path for pyspark 
(https://stackoverflow.com/questions/30983982/how-to-use-jdbc-source-to-write-and-read-data-in-pyspark)
and try to build driver with all dependencies included with 
`mvn package assembly:single -DskipTests=true`

Please tell me if it helps.

**Comment by seufagner** [comment]
Thanks for your reply.

I'm still trying. When I use Scala API I got the same error. I'll try to solve classpath problem/conflict and come back here.

**Comment by seufagner** [comment]
If I try to use driver with dependencies it does not work or assembly JAR with your dependencies as well.

To solve this, I just put some jars that conflicts with Spark version and submit my job with "extraClassPath" and point to directory where them are.

```
--conf spark.executor.extraClassPath=/$PROJECT_DIR/* --conf spark.driver.extraClassPath=/$PROJECT_DIR/*
```

JARs:
* httpclient-4.5.2.jar
* httpcore-4.4.4.jar

**Comment by Tasselmi** [comment]
> * 4.5.2

I run across the same problem.

fuck ... I can not solve it

**Comment by fibersel** [comment]
@Tasselmi  I've met this error recently and found the solution.
I think, problem is with dependencies.
Building jar from branch https://github.com/jaykelin/clickhouse-jdbc/tree/0.1-SNAPSHOT and passing them to spark-submit worked for me.I can't establish the root of this problem right now, but I hope the approach above would solve your problem.

**Comment by copy-of-rezo** [comment]
I fixed it with httpclient-4.5.4.jar and httpcore-4.4.8.jar in spark distribution jars folder
